File: algorithms/algorithms-cplusplus/leetcode/Question_0099_Recover_Binary_Search_Tree.cc

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  void recoverTree(TreeNode* root) {
    TreeNode *s = nullptr, *t = nullptr, *prev = nullptr;
    inorder(root, prev, s, t);
    int tmp = s->val;
    s->val = t->val;
    t->val = tmp;
  }

  void inorder(TreeNode* cur, TreeNode*& prev, TreeNode*& s, TreeNode*& t) {
    if(!cur) {
      return;
    }
    inorder(cur->left, prev, s, t);
    if(prev && prev->val >= cur->val) {
      s = s == nullptr ? prev : s;
      t = cur;
    }
    prev = cur;
    inorder(cur->right, prev, s, t);
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0099_Recover_Binary_Search_Tree.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0099_Recover_Binary_Search_Tree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    TreeNode pre = null, first = null, second = null;

    public void recoverTree(TreeNode root) {
        // In order traversal to find the two elements
        inorder(root);

        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }

    private void inorder(TreeNode root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        if (pre != null && first == null && pre.val >= root.val) {
            first = pre;
        }
        if (first != null && pre.val >= root.val) {
            second = root;
        }
        pre = root;
        inorder(root.right);
    }
}
